snuba/consumers/consumer.py

```diff
@@ -616,40 +639,70 @@ def process_message_multistorage(
-            .get_stream_loader()
-            .get_processor()
-            .process_message(storage_message, metadata)
+
+        result = process_message_multistorage_work(
+            metadata=metadata, storage_key=storage_key, storage_message=storage_message
         )
-        if isinstance(result, AggregateInsertBatch):
-            values_row_encoder = get_values_row_encoder(storage_key)
-            results.append(
-                (
-                    storage_key,
-                    BytesInsertBatch(
-                        [values_row_encoder.encode(row) for row in result.rows],
-                        result.origin_timestamp,
-                    ),
-                )
-            )
-        elif isinstance(result, InsertBatch):
-            results.append(
-                (
-                    storage_key,
-                    BytesInsertBatch(
-                        [json_row_encoder.encode(row) for row in result.rows],
-                        result.origin_timestamp,
-                    ),
-                )
-            )
-        else:
-            results.append((storage_key, result))
+        results.append((storage_key, result))
 
     return results
 
 
+def process_message_multistorage_identical_storages(
+    message: Message[MultistorageKafkaPayload],
+) -> Sequence[Tuple[StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]]]:
+    """
+    This method is similar to process_message_multistorage except for a minor difference.
+    It performs an optimization where it avoids processing a message multiple times if the
+    it finds that the storages on which data needs to be written are identical. This is a
+    performance optimization since we remove the message processing time completely for all
+    identical storages like errors and errors_v2.
+
+    It is possible that the storage keys in the message could be a mix of identical and
+    non-identical storages. This method takes into account that scenario as well.
+
+    The reason why this method has been created rather than modifying the existing
+    process_message_multistorage is to avoid doing a check for every message in cases
+    where there are no identical storages like metrics.
+    """
+    value = rapidjson.loads(message.payload.payload.value)
+    metadata = KafkaMessageMetadata(
+        message.offset, message.partition.index, message.timestamp
+    )
+
+    intermediate_results: MutableMapping[
+        StorageKey, Union[None, BytesInsertBatch, ReplacementBatch]
+    ] = {}
+
+    for index, storage_key in enumerate(message.payload.storage_keys):
+        skip_processing = False
+        for other_storage_key, insert_batch in intermediate_results.items():
+            if are_writes_identical(storage_key, other_storage_key):
+                intermediate_results[storage_key] = copy.copy(insert_batch)
+                skip_processing = True
+                break
+
+        if skip_processing:
+            continue
```

Comment:
Nit: I think you should be able to avoid this continue that are counter intuitive by inverting the conditions:
when you find the storage, instead of  setting `skip_processing=True` assign the message to a variable (processed_message).
Then here, if the `processed_message` is None, do the processing, otherwise store processed_message in `intermediate_results`